### PR TITLE
Added mutual cuff functionality

### PR DESCRIPTION
**Pull request by DDihanov**
[featurerequest][tested]

Middle+alt clicking on a `mob/carbon` now cuffs yourself with the mob together. It uses the `slot_handcuffed`.
It adds two events to the two mobs and when each of them move the other one slips and the former tumbles on top of the latter. When the restraints drop or the cuffs get `qdel`d the two events get deleted.

Being mutual cuffed DOES NOT impede you in the same way as being restrained e.g. you can still do everything as before, open doors interact, drop items etc, the only thing is that you cannot get handcuffed again(tethered to another person or just normal cuffed). This makes sense, as you are not really "restrained" as you can move your arms. You can still pick up items(you have a cuff icon on your arm and it overlaps with the picked up item). You can remove the other person's cuffs from their inventory if they stand still while you are doing it(like normal cuffs). You can also resist out of them. If a third party removes either of the cuffs the drop to the ground.

You cannot get in a locker if mutual cuffed, one because the logic to implement this with the locker's `forceMove` was super buggy(one person getting left outside/the second person just teleporting inside when the first one closes the door/both users slipping inside etc../the closet not even closing). So trying to close the closet while somebody who is mutually cuffed inside is not possible(you get a message). 

I tested this with syndiecuffs, wires and normal cuffs and it all works fine, since the event deleting logic is in the `Destroy()` method, and also if they drop on the ground the event remove method gets called.

Inspecting people with mutual cuffs on just shows they are handcuffed.

I also added a new sprite for the mutual cuffs which is basically the same original handcuff sprite but halved(just deleted the left part of it) so it spawns only on one arm.

EDIT: Added option for a third player to cuff two players. Drag on the first then drop on the second and they get cuffed together.

Here are some gifs for your convenience:

Cuffing somebody and moving out of range slips you and the person, tumbling him on top of you:

(with new drag)
new_cuff2

Cuff removal:

cuff_remove

Somebody else removing the cuffs:

remove_cuffs_third_party

Somebody dragging somebody that's cuffed to somebody 🤔:

cuff_drag

Monkey cuff:

cuff_monkey

Closet cuff:

locker_cuff

cuff_other_closet

Wizard blink also teles you with them, because why not, it looks fun:

wizard_blink

EDIT: 

Third player cuffing two others:

(with new drag)
cuff_new1

:cl:
 * rscadd: Added mutual cuffing. Drag yourself or somebody on top of you with the cuffs in the active hand to cuff them to yourself. You can still do normal things while being cuffed(unlike regular cuffs), with the exception of being cuffed again. If either player moves out of range - you both slip and tumble on top of each other. You cannot close a closet if there is somebody that's cuffed to somebody inside.

* rscadd: A third player can mutual cuff two others by dragging them on top of each other. Both players must stand still and be in range for this to go through.

### CODE REVIEW

**Comment by boy2mantwicethefam** [comment]
Shorten the changelog because the bot has brainfarts and long changelogs just don't show up I know this from experience

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
Needs a different shortcut than altclick

**Comment by FudgePucker** [comment]
Is there a clumbsy check when doing this? There shouldn't be.

**Comment by DDihanov** [comment]
> Is there a clumbsy check when doing this? There shouldn't be.

No clumsy check.



> Shorten the changelog because the bot has brainfarts and long changelogs just don't show up I know this from experience

Got it thanks.



> Needs a different shortcut than altclick

Well, what would you have it then? Just shift is bad cause of inspect, but shift + middle is kinda wonky no?

**Comment by RealestEstate** [comment]
alt click on a mob is technically reserved for giving. would alt middle work? I think shift middle was point.

**Comment by DDihanov** [comment]
I could try, but there would also need to be a key for a third person restraining two people together which I want to do in the future.

**Comment by optimumtact** [comment]
ERP, ERP has evolved

**Comment by DDihanov** [comment]
> ERP, ERP has evolved

If the player is degenerate enough he can ERP through anything really 🤔

**Comment by SonixApache** [comment]
*Very* cool feature. Good shit. 👍 

What about putting two cuffed people in the same locker?

I really wish you could full-cuff people that were already cuffed to someone else, would make prisoner transport really cool.

**Comment by oversane** [comment]
our get along cuffs

**Comment by DDihanov** [comment]
Added option for a third player to cuff two players. Shift+alt+middle+click on the first then  Shift+alt+middle+click on the second and they get cuffed together. Will update my main post with a gif.

**Comment by Lokiamis** [comment]
Chain gangs when?

**Comment by Blithering** [comment]
ability to handcuff briefcases to your wrist when?

**Comment by ghost** [comment]
Instead of alt-middle click why not alt-left click and alt-ctrl-middle click? Middle click has always felt clumsy to me.
Also, chang gang 2020. Add the ability to chain cuffs and it'll be the best feature in years.

**Comment by DDihanov** [comment]
> Instead of alt-middle click why not alt-left click and alt-ctrl-middle click? Middle click has always felt clumsy to me.
> Also, chang gang 2020. Add the ability to chain cuffs and it'll be the best feature in years.

Alt click gives item to player. It was alt click originally but @RealestEstate suggested alt middle. 

Also the chain cuffs may be doable but I will try it in a different update because this will become too big.

**Comment by Kitfox88** [comment]
ability to handcuff the handcuffs themselves so that you can do the magic trick where you pull apart the loops without unlocking them WHEN

**Comment by Dacendeth** [comment]
What if one person teleports?

**Comment by DDihanov** [comment]
You teleport with them.

File: code/_onclick/mutual_cuff.dm [hunk 1]
```diff
@@ -0,0 +1,158 @@
+/datum/mutual_cuff_other_players
+	var/is_selecting_other_player = FALSE
+	var/mob/living/carbon/first_player_to_cuff = null
```
Comment (ShiftyRail) [inline code/_onclick/mutual_cuff.dm]: Be certain those are properly set to null when the datum is destroyed, and make a way to set this to null when the `first_player_to_cuff` is Destroyed.

Ideally we would want an event for someone being qdel'd and have this datum listen to this event, but having worked with events in the past there's almost impossible to make sense of

File: code/_onclick/mutual_cuff.dm [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+	if(restraint_apply_sound)
+		playsound(src, restraint_apply_sound, 30, 1, -2)
+	third.visible_message("<span class='danger'>\The [third] is trying to restrain \the [first] and \the [second] together with the \the [src]!</span>",
+						 "<span class='danger'>You try to tether \the [first] and \the [second] using \the [src]!</span>")
+
+	if(do_after(third, second, restraint_apply_time))
+		//2. another range check just to be sure
+		if (!first || !third.Adjacent(first))
+			to_chat(third, "\The [first] is too far away.")
```
Comment (ShiftyRail) [inline code/_onclick/mutual_cuff.dm]: Wrap this in `warning`, please.

File: code/_onclick/mutual_cuff.dm [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+		if(first.handcuffed || first.mutual_handcuffs || second.handcuffed || second.mutual_handcuffs)
+			to_chat(third, "<span class='notice'>One of the them is already handcuffed.</span>")
+			return FALSE
+		feedback_add_details("handcuffs", "[name]")
+
+		third.visible_message("<span class='danger'>\The [third] has restrained \the [first] and \the [second] together with \the [src]!</span>")
+		third.attack_log += text("\[[time_stamp()]\] <font color='red'>Has restrained [first.name] ([first.ckey]) and the \the [second.name] ([second.key]) with \the [src](mutual cuff).</font>")
+		first.attack_log += text("\[[time_stamp()]\] <font color='red'>Restrained with \the [src] by [third.name] ([third.ckey])(mutual cuff)</font>")
+		second.attack_log += text("\[[time_stamp()]\] <font color='red'>Restrained with \the [src] by [third.name] ([third.ckey])(mutual cuff)</font>")
+		log_attack("[third.name] ([third.ckey]) has restrained [first.name] ([third.ckey]) and [second.name] ([second.ckey]) with \the [src] (mutual cuff)")
```
Comment (ShiftyRail) [inline code/_onclick/mutual_cuff.dm]: `("[third.name] ([third.ckey]) ` => `key_name(third)`

File: code/_onclick/mutual_cuff.dm [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+		var/obj/item/weapon/handcuffs/cuffs = src
+		if(istype(src, /obj/item/weapon/handcuffs/cyborg))
+			cuffs = new /obj/item/weapon/handcuffs/cyborg(get_turf(third))
```
Comment (ShiftyRail) [inline code/_onclick/mutual_cuff.dm]: Those handcuffs specifically delete themselves when dropped or uncuffed. Did you take that into account ?

File: code/_onclick/mutual_cuff.dm [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+		//important to call this AFTER setting null to the prev values
+		C.u_equip(src)
+		handcuffed_to.u_equip(src)
+
+/mob/living/carbon/proc/on_mutual_cuffed_move()
+	if (mutual_handcuffed_to && !mutual_handcuffed_to.Adjacent(src)) 
+		mutual_handcuffed_to.Slip(2, 3)
+		src.Slip(2, 3)
+		src.forceMove(mutual_handcuffed_to.loc)
+		sleep(2) //sleep as not to get too many nested calls
```
Comment (ShiftyRail) [inline code/_onclick/mutual_cuff.dm]: I'm not sure if this is a good way to go about this
Have a delayer for that, or if it's too complicated, a variable that register the last call and check if `world.time > last_call + 2`

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 6]
```diff
@@ -171,6 +171,15 @@
 	if(!src.can_close())
 		return 0
 
+	for(var/atom/movable/AM in src.loc)
```
Comment (ShiftyRail) [inline code/game/objects/structures/crates_lockers/closets.dm]: You're only looping through carbons so `for (var/mob/living/carbon/C in src.loc)` would be clearer

**Comment by DamianX** [comment]
\>shift alt middle click
there must be a better way

**Comment by ShiftyRail** [comment]
Drag one prisonner to the other with cuffs in hands ?

**Comment by DDihanov** [comment]
I thought dragging was reserved for some other stuff I didn't want to mess with, as I wanted an entirely new hotkey for it so I don't break anything. But if you have any concrete ideas I will be happy to rework it.

File: code/_onclick/mutual_cuff.dm (hunk 1 above)
Comment (DDihanov) [reply code/_onclick/mutual_cuff.dm]: You mean set the datum to null? If I use an event then whether the mob holds a new event var or just holds this datum without it being set to null is the same thing.

Comment (DDihanov) [reply code/_onclick/mutual_cuff.dm]: I tested this with cuffing one player and deleting the other and vice versa. I got no runtimes whatsoever, in the first case I got a "No other target selected" message, and the other a "Player is too far away", so it's all good.

File: code/_onclick/mutual_cuff.dm (hunk 5 above)
Comment (DDihanov) [reply code/_onclick/mutual_cuff.dm]: This was a good change, now it's smoother somehow.

File: code/game/objects/structures/crates_lockers/closets.dm (hunk 6 above)
Comment (DDihanov) [reply code/game/objects/structures/crates_lockers/closets.dm]: Good idea.

File: code/_onclick/mutual_cuff.dm (hunk 4 above)
Comment (DDihanov) [reply code/_onclick/mutual_cuff.dm]: Yeah it's all good. Syndie cuffs also do the same when destroyed, as per my main post: 

>When the restraints drop or the cuffs get qdeld the two events get deleted.

https://github.com/vgstation-coders/vgstation13/blob/9f06a64c4272bd0bd7e31fe1d1d6fc726354c9b5/code/game/objects/items/weapons/handcuffs.dm#L25-L28

**Comment by ShiftyRail** [comment]
I don't believe there's anything yet with dragging a mob to a `mob`. There is for sure dragging a mob to a table, a turf, or a locker, though.

**Comment by DDihanov** [comment]
Will look into this.

**Comment by DDihanov** [comment]
Ok so there is a problem with `MouseDrop` - it only accounts for the two mobs from where the cursor starts and ends and calls the respective MouseDropTo/From method on the mob. It doesn't pass the person who is actually holding the cuffs as an argument(a third person). If you have any other ideas then share them, but I don't see this working with drag functionality.

**Comment by ShiftyRail** [comment]
Fairly certain `usr` in the context of click code can be reliably interpreted as the person having done the action

**Comment by DDihanov** [comment]
> Fairly certain `usr` in the context of click code can be reliably interpreted as the person having done the action

This works great -  I will rework it so the normal mutual cuff is also when you drag somebody ontop of you with cuffs in hand.

**Comment by DDihanov** [comment]
Drag works great, also less code:

cuff_new1

new_cuff2

File: code/_onclick/mutual_cuff.dm (hunk 1 above)
Comment (DDihanov) [reply code/_onclick/mutual_cuff.dm]: Also I added this, which drops the cuffs when one player gets qdel'd:

https://github.com/vgstation-coders/vgstation13/blob/c87846fa519c9b39c71c1d128a4f056205da14cf/code/modules/mob/living/carbon/carbon_defines.dm#L34-L37

**Comment by PaulRitter** [comment]
Ight imma just condense my discord shitpost so people voting on this can see it:
This is a really neat addition, but the dragging logic kills it for me, it
- looks laggy
- is weirdly seperated into three different procs
- can easily be replaced by a much smoother alternative, atom locking (buckling), as @jknpj pointed out

jazz says he thinks the drag behaviour is fine and if someone wants to fix it, they can go ahead.
Imo, if you are gonna do something, don't half-arse it.

Dragging in question:
image

**Comment by DDihanov** [comment]
>looks laggy

It's not smooth as pulling logic, because it has nothing to do with pulling logic.

>can easily be replaced by a much smoother alternative, atom locking (buckling), as @jknpj pointed out
>easily
>half assed

If it's so easy then make a PR to my fork and do it. I spent a lot of time on this and made sure it works without breaking anything else. I also addressed all of Shifty's review concerns and spent a good amount of time reworking from the old alt +click(the one which I first did) to the alt click + middle click(the one that was then suggested) to the drag and drop option, until the best option was present, so no, it's not half assed.

>is weirdly seperated into three different procs

It's two procs -> one is the cuffing proc the other is the event that gets triggered.

**Comment by PaulRitter** [comment]
i was wrong, it would not be easy, though i stand behind it currently being half-assed because:
- it undeniably currently looks shitty (lets be honest here)
- you literally saying "fix it yourself" instead of fixing it

as i said in discord, this is just my opinion and since im not a c*llab its not that important, if you want to merge it at this state to upstream, go for it.
i wouldn't.

**Comment by DDihanov** [comment]
> you literally saying "fix it yourself" instead of fixing it

Don't twist my words. I said that if you imply it's so easy then you should do it. The implication on your side that it was easy was why I said it. You also said I half assed it which was wrong. I do agree it does not look smooth as pulling but it's not gamebreakingly bad IMO.

**Comment by PaulRitter** [comment]
image
im not twisting your words, don't get butthurt, this isn't personal, i just don't think this can be considered ready to merge

**Comment by DDihanov** [comment]
Yeah but you left out the part earlier where you said it's easy. Don't just cherry pick comments to paste. You are turning this into a real social network drama.

> don't get butthurt, this isn't personal

You are the one making it personal but representing my arguments to fit your cause, cherry picking comments, and saying I half assed it without even looking at the commits. I just called you out on everything you were wrong about, and now you are calling me butthurt, while you are the one getting defensive and posting screenshots to prove your point. If you don't like the feature just downvote it and be done with it.

**Comment by PaulRitter** [comment]
Don't misunderstand me, this is a cool feature, as well as the possibility of chain cuffing in the near (or not, doesn't matter) future.

i'm not trying to fight you, commits and the code don't matter for this argument, even me saying its easy to change which i admitted i wrongfully claimed.

All this doesn't change you saying that we should change it ourselves as well as my opinion that this is a bad behaviour to approach critizism

My opinion about the this being ready to merge won't change unless dragging gets changed.

Edit: I will not just downvote this without stating why, overread that part

**Comment by DDihanov** [comment]
Let me explain something about the dragging. When somebody is dragging a pair of mutually cuffed players, what is happening is the same logic as in the other cases. 

For the sake of example let's say person A is dragging person B who is mutually cuffed to C.

If B gets out of range of C, B and C are slipped and tumbled on top of each other. What is happening in this example is the same thing -> one of the players gets out of range of the other(because A is dragging him out of range) and that's why there is the constant teleporting. If both players(B and C) were to move synchronously and stay in range - this wouldn't happen. So if A is dragging C and B was to move alongside C this wouldn't be happening. There is no way to make this "smooth" as the core functionality is still the same.

**Review by ShiftyRail — APPROVED** [review]

**Comment by ShiftyRail** [comment]
Can't see anything that stands out as of now

File: code/_onclick/click.dm [hunk 7]
```diff
@@ -32,6 +32,18 @@
...
+/mob/living/carbon/MouseDrop(var/mob/living/carbon/first, var/second_turf, over_location, src_control, over_control, params)
+	var/mob/living/carbon/second = locate() in second_turf
+	if (!istype(first) || !second || (first == usr && second == usr) || (first == second)) //if user is dragging only on himself or user drags and drops on the same target
+		return
+	var/obj/item/to_be_handcuffs = usr.get_active_hand()
+	if (first.Adjacent(usr) && second.Adjacent(usr) && istype(to_be_handcuffs, /obj/item/weapon/handcuffs))
+		var/obj/item/weapon/handcuffs/handcuffs = to_be_handcuffs
+		handcuffs.apply_mutual_cuffs(first, second, usr)
+		return
+	..()
```
Comment (ShiftyRail) [inline code/_onclick/click.dm:45]: Probably should have been `return ..()`

**Comment by bathosbathtime** [comment]
Seems to have borked ghetto cuffing (see #23416 and #23397)

**Pull request merged**